[pluiexo] iP

Pluiexo opened the pull request:
# DukePro Max
>  "History repeats itself, first as tragedy, second as farce."- Karl marx

Duke frees your mind and it will enlighten you to live a fulfilling life. Its
*  Sentient
*  ~~Small~~ Super Small Package 
*  Smart and easy to use 

All you need to do is,
1. Run it from [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ)
2. Double click the link above
3. Add your task!
4. Done and let the magic happen! 🎆 

And its **FREE!!!!**

Features:

- [x] Adding task
- [x] Managing task
- [ ]  Sell your tasks (Coming soon!)

---

```java
public class Duke {
    public static void main(String[] args) {
        Application.launch(DukeLauncher.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Level-0

Commit:
Level 1

Commit:
Level-2

Commit:
Level-3

Commit:
Level-4

Commit:
Fix bugs and added TextUiTesting

Commit:
Tidy up classes and added enums

Commit:
Level-5

Commit:
Level-6

Commit:
Updated Level-6

Commit:
Add save functionalities

Commit:
Fix bugs in save system

Commit:
Fix and tidy save system

Commit:
Fix and tidy save system

Commit:
Merge branch 'branch-level-7'

Commit:
Add DateHandler to mananage DateTime in String format

Commit:
Add regex to parse inputs

Commit:
Add DateTime objects to events classes

Commit:
Merge branch 'branch-level-8'

Commit:
Reorganise into better OOP classes

Commit:
Fix broken methods in Date Handler

Commit:
Merge branch 'A-MoreOOP'

Commit:
Reorganise into packages

Commit:
Merge branch 'branch-A-Packages'

Commit:
Refactor code into better parts

Commit:
Merge branch 'branch-add-gradle-support'

Commit:
Add JUnit tests

Commit:
Merge branch 'branch-A-JUnit'

Commit:
Tidy up formatting issues

Commit:
Build jar with gradle

Commit:
Change DateHandler date to support integer input

Commit:
Add Java Docs

Commit:
Tidy code up

Commit:
Correct typos in Java Doc

Commit:
Add find function to search for tasks

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-Level-9'

Commit:
Reformat code

xinen26 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,137 @@
+        case "D":
+            name = data[2];
+            by = data[3];
+            String temp = data[4];
```
just a thought but maybe can use a different variable name instead of "temp" to provide more info and make things clearer :)

xinen26 commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,252 @@
+ * Manager class to manage and keep track of all the actions being performed on the task.
+ */
+public class TaskManager {
+    private static final String listingResponse = "Here are the tasks in your list:";
```
just a suggestion but maybe can change this string name to RESPONSE_LIST

xinen26 commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,252 @@
+        String description;
+        String by;
+        String from;
+        Pattern deadlineFormat = Pattern.compile("(?<description>.+)\\s?((?i)/by)(?<by>.+)");
```
wow, this is really impressive! it never crossed my mind to implement it, but after seeing your code, i'm definitely going to give it a try.

xinen26 commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,252 @@
+            } else if (description.isBlank()) {
+                throw new DukeException("description");
+            }
+            Optional<LocalDate> testByDate = DateHandler.checkDate(by);
```
i would have never thought of using Optional and FlatMap and i think it is quite cool that you managed to incorporate it into your IP! 💯

xinen26 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,135 @@
+        //Used when your user does not know how to use the application
+        String errorMsg;
+        switch (error) {
+        case "description":
```
a nitpick i have is that maybe you can try to further standardise their names. for eg: dateError has Error behind it whereas no other exceptions has Error behind their name. Maybe you can try to include it so for eg invalidInputError, missingIndexError and etc. Also, for your DateOutOfRange, i realised that none of the other errors have their first letter capitalised so maybe you can consider changing it if you want

xinen26 replied:
just a tiny suggestion but honestly it looks really good with or without the change 👍

xinen26 commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,135 @@
+        String errorMsg;
+        switch (error) {
+        case "description":
+            errorMsg = " OOPS!!! The description of this command cannot be empty.";
```
in my personal opinion, maybe can change to errorMessage instead of errorMsg to make it even clearer but i am pretty sure errorMsg works too

xinen26 reviewed (commented):
truly enjoyed how you've integrated Optionals, Flatmaps, and other techniques into your code though it was not mentioned in this module! reviewing it has sparked numerous ideas on how i can further enhance my own code :)

Commit:
Add checkstyle

Commit:
Fix style issues

Commit:
Merge branch 'branch-A-CheckStyle'

Commit:
Add GUI

Commit:
Tidy code up

Commit:
Fix spacing in dialog label

Commit:
Merge branch 'branch-Level-10'

Commit:
Change arraylist to vaargs

Commit:
Fix Junit test

Commit:
Merge branch 'branch-A-Varags'

Commit:
Fix JavaDocs

yuhinarita reviewed (commented):
~~Overall, looks very good. Just that I thought your switch expression might not be supported in Java 11~~
Sorry please ignore this, it was for someone else's code.

yuhinarita commented on src/main/java/duke/parser/DateHandler.java:
```diff
@@ -0,0 +1,100 @@
+    private static final Pattern PATTERN_DATE = Pattern.compile(
+            "([a-zA-Z]+)?\\s?(?<d>\\d{2})[-/](?<m>\\d{1,2})[-/](?<y>\\d{2,4})\\s?([a-zA-Z]+)?");
+    private static final Pattern TIME_PATTERN = Pattern.compile(
+            "(.+)?\\s?(?<time>(\\d{2}:?\\d{2}))(?<indicator>(?i)\\s?[ap]m)?\\s?(.+)?");
```
I really like your use of regular expression! Makes your parser robust and flexible.

yuhinarita commented on src/main/java/duke/task/TaskManager.java:
```diff
@@ -0,0 +1,264 @@
+            }
+            Optional<LocalDate> testDate = DateHandler.checkDate(by);
+            item = testDate
+                    .map(localDate -> new Deadline(description,
```
I like how your code is written with lots of functional programming concepts in mind. I did not think of using Optional, but looking at this, it's definitely something I would want to consider now.

yuhinarita reviewed (commented):
Overall, the code is clean, well designed, and documented, with nothing much to comment further from my end. I actually learned more reviewing your code, such as use of Pattern & regexp, or how to incorporate Optional and functional programming elements. Keep up the good work!

Commit:
Add assertions to test important assumptions

There are seveal methods that assumes that the argument passed in is
valid. In general usage this is true, however there may be cases where
that is forgetten and an invalid value may be passed into the function.
This can cause the program to perform erroneously to continue.

As a fail safe for such a catastrophic event, let's add several
assertion to indicate to us what had gone wrong.

Pluiexo committed:
Merge pull request #1 from Pluiexo/branch-A-Assertions

Add assertions to test important assumptions

Commit:
Refactor TaskManager

Commit:
Fix TaskManager bugs

Commit:
Refactor storage class

Commit:
Refactor and tidy code

Commit:
Tidy up TaskManager

Commit:
Add Stream functions to finding task

Pluiexo committed:
Merge pull request #2 from Pluiexo/branch-A-Streams

Add Stream functions to finding task

Commit:
Fix merge conflicts

Pluiexo committed:
Merge pull request #3 from Pluiexo/branch-A-CodeQuality

Branch a code quality

Commit:
Add github actions

Commit:
Add query functions

Commit:
Add view task by schedule function

Commit:
Add clear function to clear the chatbot

Commit:
Fix bugs

Commit:
Merge branch 'branch-BCD-Extension'

Commit:
Improve view schedule function

Commit:
Change index system for query functions

Commit:
Fix formatting issues

Commit:
Reorder statements

Commit:
Fix JavaDocs

Commit:
Edit Images to fit better

Commit:
Add new UI enchancements

Commit:
Fix list response

Commit:
Merge branch 'branch-A-BetterGui'

Commit:
Tweak values for dropshadow

Commit:
Add a new personality Plato

Commit:
Fix formatting issues

Commit:
Merge branch 'branch-A-Personality'

Commit:
Add more error handling to the time function in DateHandler

Commit:
Tidy TaskManager

Commit:
Merge branch 'branch-A-MoreErrorHandling'

Commit:
Fix colors

Commit:
Add more Junit tests

Commit:
Merge branch 'branch-A-MoreTesting'

Commit:
Update README in docs

Commit:
Correct dateHandler format issues

Commit:
Update README

Commit:
Update README

Commit:
Update Readme

Commit:
Update README

Commit:
Merge branch 'branch-A-UserGuide'

Commit:
Fix additional classes

Commit:
Tidy JavaDocs

Commit:
Fix gradle

Commit:
Fix gitignore

Commit:
Fix parser tests
